src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs

```diff
@@ -29,7 +29,8 @@ public static void Start(this IHost host)
         /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
         public static Task StopAsync(this IHost host, TimeSpan timeout)
         {
-            return host.StopAsync(new CancellationTokenSource(timeout).Token);
+            using CancellationTokenSource cts = new CancellationTokenSource(timeout);
+            return host.StopAsync(cts.Token);
```

Comment:
I was getting a warning in this code saying that CancellationTokenSource was not being disposed, so I added this. Let me know if it was not the right thing to do.

Reply:
This is incorrect. This code will immediately dispose of the CTS right after calling StopAsync, even before StopAsync completes (since you're not awaiting it). as a result, you're rendering the Timeout useless because it gets disposed before it ever expires.
The appropriate thing to do here is dispose of the CTS after awaiting the StopAsync call. That way if StopAsync completes quickly, the native timer resource can be released, whereas it was just getting abandoned before and only released after the timeout actually expired.

Reply:
Looks fine.

Reply:
Reverted.

Reply:
It'd be even better to change this method to dispose the CTS when StopAsync completes as @AArnott suggested.

```C#
        public static async Task StopAsync(this IHost host, TimeSpan timeout)
        {
            using CancellationTokenSource cts = new CancellationTokenSource(timeout);
            await host.StopAsync(cts.Token);
        }
```

Reply:
This is a public method. Wouldn't it be a breaking change if the method is made `async`?

Reply:
I don't see how it would be a breaking change. *async* doesn't manifest in the signature, and the behavior isn't changing because it was already returning a Task that may or may not be Completed by the time the method returns. 
But you should add `.ConfigureAwait(false)` to the awaited expression so you don't inadvertently introduce a dependency on some SynchronizationContext.

Reply:
Thanks for the explanation. Updated.